From d033f717faaa1c41f4d80f564e21afe66583f7ed Mon Sep 17 00:00:00 2001
From: Alexandru Costache <alexandru@balena.io>
Date: Wed, 9 Mar 2022 15:05:48 +0100
Subject: [PATCH] The current code invalidates the range after the read buffer
 since the buffer pointer gets incremented in the read loop. Use a temporary
 pointer to make sure we have a pristine pointer to invalidate the correct
 memory range after read.

Fixes: 704e040a51d2 ("nvme: Apply cache operations on the DMA buffers")

Patch source: https://lists.denx.de/pipermail/u-boot/2021-September/461882.html

Upstream-Status: Backport
Signed-off-by: Stefan Agner <stefan at agner.ch>
Signed-off-by: Alexandru Costache <alexandru@balena.io>
---
 drivers/nvme/nvme.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/nvme/nvme.c b/drivers/nvme/nvme.c
index 0b7082d71b..97b533f85c 100644
--- a/drivers/nvme/nvme.c
+++ b/drivers/nvme/nvme.c
@@ -746,6 +746,7 @@ static ulong nvme_blk_rw(struct udevice *udev, lbaint_t blknr,
 	u64 prp2;
 	u64 total_len = blkcnt << desc->log2blksz;
 	u64 temp_len = total_len;
+	void *temp_buffer = buffer;
 
 	u64 slba = blknr;
 	u16 lbas = 1 << (dev->max_transfer_shift - ns->lba_shift);
@@ -773,19 +774,19 @@ static ulong nvme_blk_rw(struct udevice *udev, lbaint_t blknr,
 		}
 
 		if (nvme_setup_prps(dev, &prp2,
-				    lbas << ns->lba_shift, buffer))
+				    lbas << ns->lba_shift, (ulong)temp_buffer))
 			return -EIO;
 		c.rw.slba = cpu_to_le64(slba);
 		slba += lbas;
 		c.rw.length = cpu_to_le16(lbas - 1);
-		c.rw.prp1 = cpu_to_le64(nvme_virt_to_bus(dev, buffer));
+		c.rw.prp1 = cpu_to_le64(nvme_virt_to_bus(dev, (ulong)temp_buffer));
 		c.rw.prp2 = cpu_to_le64(prp2);
 		status = nvme_submit_sync_cmd(dev->queues[NVME_IO_Q],
 				&c, NULL, IO_TIMEOUT);
 		if (status)
 			break;
 		temp_len -= (u32)lbas << ns->lba_shift;
-		buffer += lbas << ns->lba_shift;
+		temp_buffer += lbas << ns->lba_shift;
 	}
 
 	if (read)
-- 
2.17.1

